### PR TITLE
Added Wool to OreDictionary

### PR DESCRIPTION
**Pull request by lclc98**
Added wool to ore dictionary. I have tested it with beds to make sure it didn't change vanilla behavior with color specific recipes.

### CODE REVIEW

**Review by Darkhax — APPROVED** [review]
This change would be useful for many mods. Newer versions of the game include wool in their tag system, and I can think of several existing and unreleased mods on 1.12.2 that would greatly benefit from this.

**Review by tterrag1098 — APPROVED** [review]

File: src/main/java/net/minecraftforge/oredict/OreDictionary.java [hunk 1]
```diff
@@ -226,6 +226,9 @@ private static void initVanillaEntries()
             registerOre("paneGlass",     Blocks.GLASS_PANE);
             registerOre("paneGlass",     new ItemStack(Blocks.STAINED_GLASS_PANE, 1, WILDCARD_VALUE));
             //paneGlass{Color} is added below with dyes
+            registerOre("wool",          Blocks.WOOL);
```
Comment (bs2609) [inline src/main/java/net/minecraftforge/oredict/OreDictionary.java]: This line shouldn't be here - all wool is "dyed" wool, there's no seperate "clear" wool block like glass has.

Comment (lclc98) [reply src/main/java/net/minecraftforge/oredict/OreDictionary.java]: Thanks for pointing that out.

**Comment by KnightMiner** [comment]
Shouldnt there be a replacement for wildcard wool for the sake of the painting recipe?

**Comment by lclc98** [comment]
Yeah, you are right, fixing that now. Thanks for pointing it out

**Comment by bs2609** [comment]
> Shouldnt there be a replacement for wildcard wool for the sake of the painting recipe?

I don't think that will work, actually.

The vanilla painting recipe just (in 1.12) lists the 16 available wool colours as seperate sub-ingredients.

The ingredient replacing code is strict:
https://github.com/MinecraftForge/MinecraftForge/blob/d5ff207fabd9d50c8739c6df5dc98d8ede7ea562/src/main/java/net/minecraftforge/oredict/OreDictionary.java#L414

which means it expects an exact meta match, without wildcarding:
https://github.com/MinecraftForge/MinecraftForge/blob/d5ff207fabd9d50c8739c6df5dc98d8ede7ea562/src/main/java/net/minecraftforge/oredict/OreDictionary.java#L634-L641

For it to match here, all the regular metadata values would have to be added as replacements as well, like for wood planks:
https://github.com/MinecraftForge/MinecraftForge/blob/d5ff207fabd9d50c8739c6df5dc98d8ede7ea562/src/main/java/net/minecraftforge/oredict/OreDictionary.java#L244-L250

However this would cause them to be rejected, as they would then also map to the oredict names for each colour of wool:
https://github.com/MinecraftForge/MinecraftForge/blob/d5ff207fabd9d50c8739c6df5dc98d8ede7ea562/src/main/java/net/minecraftforge/oredict/OreDictionary.java#L417-L422

Not sure what a good solution would be here, really.

**Comment by bs2609** [comment]
On a related note, this means that the slab replacement here doesn't actually work currently:
https://github.com/MinecraftForge/MinecraftForge/blob/d5ff207fabd9d50c8739c6df5dc98d8ede7ea562/src/main/java/net/minecraftforge/oredict/OreDictionary.java#L251

(this can be seen with the daylight detector recipe)

**Comment by bs2609** [comment]
Also linking #4310 for reference.

**Comment by GirafiStudios** [comment]
@bs2609 doing it like that would mean you would have to add recipes like the banner recipes to the exclusion list, since it would make any banner recipe accept any combination of colors.
Adding it to the exclusion list would mean the banner recipe would not accept non-vanilla dye-specific wool.

**Comment by bs2609** [comment]
Adding exclusions to go along with the specific meta values would be required, like the wood handling in the linked PR. Sorry I didn't mention that in the first post, it was supposed to be there, but apparently I overlooked it during the writeup.

However the point here is that it won't work anyway - the issue with adding exclusions just furthers that.

**Comment by GirafiStudios** [comment]
Yea I agree. Not adding the wool wildcard to replacements is probably the best way.

**Review by GirafiStudios — APPROVED** [review]

**Comment by tterrag1098** [comment]
I would like some real-world tests for this. Perhaps pull in Thermal mods (with their rockwool equivalents) and make sure recipes continue to function as expected.

**Comment by tterrag1098** [comment]
@GirafiStudios provided me with some such tests.

**Review by mezz — APPROVED** [review]

**Pull request merged**